### ISSUE TITLE
Entering test vector in HEX format

### ISSUE THREAD

**Issue by Elliot1981** [bauhuasbadguy/Twofish_encryption#1]
Good afternoon, dear respective bauhuasbadguy!

If is it possible could you please say how can I enter test vector in hexadecimal format at the main part of your implementation? I would be grateful for your reply! So I guess I need to use functions encrypt_word and decrypt_word I have entered particular this test vector in that format:

#set the key
key = '\x9F\x58\x9F\x5C\xF6\x12\x2C\x32\xB6\xBF\xEC\x2F\x2A\xE8\xC3\x5A'
N = 128
rounds = 16

[K, S] = gen_keys(key, N, rounds)

#there's a wierd bug where if the number of characters is a certan amount a random
#set of characters appear at the beginning. It's really strange

test = '\xD4\x91\xDB\x16\xE7\xB1\xC3\x9E\x86\xCB\x08\x6B\x78\x9F\x54\x19'

[num_C, Cypher_text] = encrypt_word(test, S, K)

print(Cypher_text)

plain_text = decrypt_word(num_C, S, K, rounds = 16)

print(plain_text)

I=3
KEY=9F589F5CF6122C32B6BFEC2F2AE8C35A
PT=D491DB16E7B1C39E86CB086B789F5419
CT=019F9809DE1711858FAAC3A3BA20FBC3

However I got the error "line 940, in <module>
    [num_C, Cypher_text] = encrypt_word(test, S, K); line 731, in encrypt_word
    E[i] = pad_number(int(e, 2) ^ K[i], 32)
ValueError: invalid literal for int() with base 2:"

Could you please say how is it possible to enter both 128 BIT KEY and 128 BIT PLAINTEXT in HEX format? Thank you in advance for your reply!

**Comment by bauhuasbadguy** [comment]
Hi Elliot1981,

I'm afraid the code isn't particularly well written so it can only take in strings. There's a couple of things going on here so the response will be a bit long winded.

The first issue is that python interprets things like \x58 as a control character and therefore when it tries to do anything with it it gets confused. You might see this referred to in my other codes as 'the characters bug' and its why when you look at how I've output a cyphertext which is intended to be put back into the decryption function it's as a base ten integer. The cypher text you see outputted is just for show and is not used by the code to prevent confusion.

To fix the problem with control characters you might want to change the way you're defining hex integers. From what I understand you need to define them as:-
key = '0x9f589f5f6122c32b6bfec2f2ae8c35a'
test = '0xD491DB16E7B1C39E86CB086B789F5419'.
(upper or lower case your choice)

Now on to the meat of the issue, the reason for the crashes.

I've done a horrible job of coding the blocking procedure as you can tell from the addition of random characters to messages of lengths that aren't directly divisible into 128 bit chunks. That particular piece of horribleness is not what's causing this particular bugbear however. What is causing it is that if the string is less than 128 bits long the chucking section thinks each individual bit is a separate block. A fix for this is coming very soon to the main code so thanks for pointing it out :)

In summary the things you need to do to make this work with hex numbers are as follows:-
1. Everywhere the code converts from strings to numbers make it convert from hexadecimal to decimal
2. Add a bit into the blocking procedure so that it puts single blocks into lists (coming soon to the main code)
3. Have tea

That should sort you out. I've tried it in a rough and ready sort of way on my laptop and it seems to work ok. In a real implementation you'd want to use hex for this sort of thing as you've suggested however I've used strings since this code was written as an educational aid to myself and to others. To that end I hope you've found it useful.

Cheers,

Bauhaus_bad_guy

**Comment by Elliot1981** [comment]
Good afternoon, dear respective bauhuasbadguy!

Thank you very much for a very detailed reply and your recommendations! I am currently implementing Twofish-128 too but got stuck on SBOX and I have sorted out the GF2^8 polynomial. Could you please recommend any reference implementations of TWOFISH-128 with main function which takes both PT, KEY in HEX format I recently looked on reference implementations on https://www.schneier.com/academic/twofish/download.html however they are programmed in API code style with out main function, so understand and identify variables output is complicated? Or maybe you have met any papers with working example of Encrypting Twofish-128 block with input and outputs of SBOXES and rounds? Thank you!

**Comment by bauhuasbadguy** [comment]
Hi Elliot1981,

I used the original paper on twofish to get the s-boxes:- 
https://www.certainkey.com/resources/article/twofish.pdf

There's also some example encryptions at the end of the paper but I didn't test that they worked as planned since I stopped working on twofish when I realised I could get my original input back when I decrypted the message. This means there's a good chance there's still a bug somewhere in this implementation somewhere but I wasn't sure how to easily find it since the examples aren't step by step so I called it a day and moved on to my AES implementation. If you find a worked sample that gives the expected results at each step could you let me know so I can check my code is doing what it should.

Thanks
Bauhaus_bad_guy

**Comment by bauhuasbadguy** [comment]
Hi again Ellitot1981,

Sorry for leaving this so long but I've decided lockdown is the time to get round to fixing up all this old code and trying to document it all. I understand that this is way way too late but having just gone through the code I figured I could share a few insights I got whilst going through the code in case it helps anyone else.

1. The s-boxes:- Whist it seems like the paper states these are substitution boxes I think they are actually h-functions taking in the s-vector generated in the key generation step as the vector input.
2. The weird extra characters at the start of the decrypted text bug:- This was being caused by a problem with the way I was padding strings of bits. I've fixed it now
3. 192 and 256 bit keys not working:- This was caused by another stupid bug and is now fixed. I should probably add in an exception for when the user provides a key that is too long but I'm going to leave it as is for now since its more stable and I want to move on
4. Other bug fixes:- In the original code I was doing the PHT wrong as well as the q boxes so that implementation did indeed have a number of major errors. These have been fixed but I don't know if there are more lurking in the code somewhere.

Sorry for the 4 year delay
Bauhaus_bad_guy
